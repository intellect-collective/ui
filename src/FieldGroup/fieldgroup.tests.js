import React from 'react';
import { FieldGroup, Text } from '../..';

describe('Form Components - FieldGroup', () => {
    it('Can render with string wrapper', () => {
        const wrapper = mount(
            <FieldGroup name="test" wrapper="div">
                <Text />
            </FieldGroup>
        );
        expect(wrapper.find('div').length).toEqual(1);
        expect(wrapper.find('div').props()).toMatchObject({
            role: 'group'
        });
    });
});
