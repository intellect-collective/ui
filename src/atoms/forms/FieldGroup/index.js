import React from 'react';
import PropTypes from 'prop-types';
import { groupContext, groupValue } from '../prop-types';

const getValue = (value) => {
    if (value === null || value === undefined) {
        return [];
    }
    if (Array.isArray(value)) {
        return [...value];
    }
    return [value];
};

export default class FieldGroup extends React.Component {
    static propTypes = {
        name: PropTypes.string,
        multiple: PropTypes.bool,
        value: groupValue,
        onChange: PropTypes.func,
        children: PropTypes.node.isRequired,
        component: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.func,
            PropTypes.object
        ])
    };

    static defaultProps = {
        component: 'div'
    };

    static childContextTypes = {
        group: groupContext
    };

    constructor() {
        super();
        this.onChange = this.onChange.bind(this);
    }

    getChildContext() {
        const { name, value } = this.props;
        return {
            group: {
                name,
                value,
                onChange: this.onChange
            }
        };
    }

    onChange(name, value, ev) {
        const { checked } = ev.target;

        // If singular
        if (!this.props.multiple) {
            this.props.onChange(ev);
            return;
        }

        // Otherwise multiple values allowed
        const val = getValue(this.props.value);
        if (checked) {
            val.push(value);
        } else {
            let index = val.indexOf(value);
            while (index !== -1) {
                val.splice(index, 1);
                index = val.indexOf(value);
            }
        }
        this.props.onChange(ev);
    }

    render() {
        const {
            name, value, onChange, component: Component, children, ...rest
        } = this.props;
        return (
            <Component role="group" { ...rest }>
                { children }
            </Component>
        );
    }
}
