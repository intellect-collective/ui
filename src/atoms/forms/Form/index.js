import React from 'react';
import PropTypes from 'prop-types';

export const FormContext = React.createContext(null);

export default class Form extends React.Component {
    static propTypes = {
        action: PropTypes.string.isRequired,
        method: PropTypes.oneOf(['get', 'post', 'put']),
        overrideProperty: PropTypes.string,
        encType: PropTypes.oneOf(['text', 'multipart']),
        onChange: PropTypes.func,
        onSubmit: PropTypes.func,
        children: PropTypes.any,
        data: PropTypes.object
    };

    static defaultProps = {
        method: 'post',
        overrideProperty: '_method',
        data: {},
        onChange: () => {},
        onSubmit: () => {}
    };

    constructor(props) {
        super(props);
        this.getValue = this.getValue.bind(this);
        this.setValue = this.setValue.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.getEncType = this.getEncType.bind(this);
        this.getFormMethod = this.getFormMethod.bind(this);
        this.getMethodOverrideElement = this.getMethodOverrideElement.bind(this);
    }


    getValue(name) {
        if (this.props.data) {
            return this.props.data[name];
        }
        return undefined;
    }

    setValue(ev) {
        this.props.onChange(ev);
    }

    /**
     * Calls the onSubmit prop function, if provided, when the form is
     * submitted.
     * @param {Object} e - The React event
     */
    onSubmit(e) {
        if (this.props.onSubmit) {
            this.props.onSubmit(this.props.data, e);
        }
    }

    /**
     * Determines the encoding type of the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The mime type to specify in th enctype attribute
     */
    getEncType() {
        if (this.props.encType === 'text') {
            return 'text/plain';
        }
        if (this.props.encType === 'multipart') {
            return 'multipart/form-data';
        }
        return this.props.encType;
    }

    /**
     * Determines the HTTP method to use when submitting the form.
     * See https://www.ietf.org/rfc/rfc1867.txt
     * @returns {String} - The http method to use
     */
    getFormMethod() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return this.props.method;
        }
        return 'post';
    }

    /**
     * Produces a hidden element which provides the opportunity to emulate
     * http methods other than GET and POST. Requires that the server receiving
     * the request support method overriding.
     * @returns {Object} - A React element
     */
    getMethodOverrideElement() {
        if (['get', 'post'].indexOf(this.props.method) !== -1) {
            return null;
        }
        return (
            <input type="hidden"
                    name={ this.props.overrideProperty || '_method' }
                    value={ this.props.method } />
        );
    }

    /**
     * Builds the props that are to be passed through to the form element. We
     * need to get rid of some props that aren't supported by the form element,
     * so we take that opportunity here.
     * @returns {Object} - The final props object
     */
    getProps() {
        const props = {
            ...this.props,
            encType: this.getEncType(),
            method: this.getFormMethod(),
            onSubmit: this.onSubmit
        };
        delete props.overrideProperty;
        return props;
    }

    /**
     * Perform render
     * @returns {Object} - A React element
     */
    render() {
        const {
            children,
            onChange,
            onSubmit,
            ...props
        } = this.getProps();
        const ctx = {
            getValue: this.getValue,
            setValue: this.setValue
        };
        return (
            <form { ...props }>
                <FormContext.Provider value={ ctx }>
                    { this.getMethodOverrideElement() }
                    { this.props.children }
                </FormContext.Provider>
            </form>
        );
    }
}
