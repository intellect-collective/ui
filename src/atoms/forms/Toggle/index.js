import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {
    asCheckable, asField, asGroupable, compose
} from '../utils';

const classes = (className, oval, worded) => (
    classnames(
        className || 'toggle',
        {
            'toggle-oval': oval,
            'toggle-worded': worded
        }
    )
);

class Toggle extends React.PureComponent {
    static propTypes = {
        name: PropTypes.string.isRequired,
        id: PropTypes.string,
        className: PropTypes.string,
        checked: PropTypes.bool,
        oval: PropTypes.bool,
        worded: PropTypes.bool
    };

    render() {
        const {
            className,
            oval,
            worded,
            ...rest
        } = this.props;
        return (
            <div className={ classes(className, oval, worded) }>
                <input type="checkbox" tabIndex="0" { ...rest } />
                <label htmlFor={ this.props.id || this.props.name }>
                    <span className="switch"><span className="switch-inner" /></span>
                </label>
            </div>
        );
    }
}

export default compose(
    asGroupable(),
    asField(),
    asCheckable()
)(Toggle);
