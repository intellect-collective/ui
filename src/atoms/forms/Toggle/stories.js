import React from 'react';
import { storiesOf } from '@storybook/react';
import { Toggle } from '../../../..';
import width from '../../../../.storybook/decorators/width';

storiesOf('Atoms / Forms / Toggle', module)
    .addDecorator(width(200))
    .add('default', () => (
        <div>
            <Toggle name="first" />
            <Toggle className="toggle toggle-success" name="second" />
            <Toggle className="toggle toggle-info" name="third" />
            <Toggle className="toggle toggle-warning" name="fourth" />
            <Toggle className="toggle toggle-danger" name="fifth" />
        </div>
    ))
    .add('worded', () => (
        <div>
            <Toggle worded name="first" />
            <Toggle worded className="toggle toggle-success" name="second" />
            <Toggle worded className="toggle toggle-info" name="third" />
            <Toggle worded className="toggle toggle-warning" name="fourth" />
            <Toggle worded className="toggle toggle-danger" name="fifth" />
        </div>
    ))
    .add('oval', () => (
        <div>
            <Toggle oval name="first" />
            <Toggle oval className="toggle toggle-success" name="second" />
            <Toggle oval className="toggle toggle-info" name="third" />
            <Toggle oval className="toggle toggle-warning" name="fourth" />
            <Toggle oval className="toggle toggle-danger" name="fifth" />
        </div>
    ))
    .add('oval worded', () => (
        <div>
            <Toggle oval worded name="first" />
            <Toggle oval worded className="toggle toggle-success" name="second" />
            <Toggle oval worded className="toggle toggle-info" name="third" />
            <Toggle oval worded className="toggle toggle-warning" name="fourth" />
            <Toggle oval worded className="toggle toggle-danger" name="fifth" />
        </div>
    ));
