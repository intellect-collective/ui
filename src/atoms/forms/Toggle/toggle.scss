@import 'mixins/toggle-size';
@import 'mixins/toggle-variant';

$toggle-font-weight:           normal !default;
$toggle-default-color:         #333 !default;

$toggle-border-radius-base:    $border-radius-base !default;
$toggle-border-radius-lg:      $border-radius-lg !default;
$toggle-border-radius-sm:      $border-radius-sm !default;

$toggle-sizes: (
    '.toggle': (52px, 20px, 20px)
) !default;

.toggle {
    position: relative;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;

    input[type="checkbox"] {
        opacity: 0;
        position: absolute;

        &:checked {
            + label .switch-inner {
                margin-left: 0;
            }
            + label .switch {
                right: 0;
            }
        }
    }

    label {
        display: block;
        overflow: hidden;
        cursor: pointer;
        border-radius: 2px;
    }

    .switch {
        display: block;
        position: relative;
        top: -2px;
        transition: right 0.3s ease-in 0s;

        &-inner,
        &:before,
        &:after {
            box-sizing: border-box;
            display: inline-block;
            font-family: Trebuchet, Arial, sans-serif;
            font-size: 12px;
            font-weight: bold;
            margin: 0;
            padding: 0;
            vertical-align: top;
        }
        &-inner {
            border-width: 2px;
            border-style: solid;
        }

        &:before,
        &:after {
            content: " ";
            text-align: center;
        }
    }

    &-oval {
        label {
            border-radius: 100px;
        }
        .switch-inner {
            border-radius: 100px;
        }
    }

    &-worded {
        .switch {
            content: " ";
            &:before {
                content: "ON";
            }
            &:after {
                content: "OFF";
            }
        }
    }

    @include toggle-variant(#999);
}

@each $class, $attributes in $toggle-sizes {
    #{$class} {
        @include toggle-size($attributes...);
    }
}

@each $color, $value in $colors {
    .toggle-#{$color} {
        @include toggle-variant($value);
    }
}