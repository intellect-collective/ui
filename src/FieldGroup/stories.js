import React from 'react';
import { storiesOf } from '@storybook/react';
import { withState } from '@dump247/storybook-state';
import FieldGroup from './index';
import Checkbox from '../Checkbox';
import Radio from '../Radio';

storiesOf('FieldGroup', module)
    .add('radio buttons', withState({ initialState: {} })(({ store }) => (
        <FieldGroup name="test" value={ store.state.test } onChange={ (ev) => store.set({ [ev.target.name]: ev.target.value }) }>
            First
            {' '}
            <Radio value="first" />
            <br />
            Second
            {' '}
            <Radio value="second" />
            <br />
            Third
            {' '}
            <Radio value="third" />
        </FieldGroup>
    )))
    .add('checkboxes', withState({})(({ store }) => (
        <FieldGroup name="test" value={ store.state.test } multiple onChange={ store.set }>
            First
            {' '}
            <Checkbox value="first" />
            <br />
            Second
            {' '}
            <Checkbox value="second" />
            <br />
            Third
            {' '}
            <Checkbox value="third" />
        </FieldGroup>
    )));
