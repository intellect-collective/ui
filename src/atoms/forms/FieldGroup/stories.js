import React from 'react';
import { storiesOf } from '@storybook/react';
import FieldGroup from './index';
import Checkbox from '../Checkbox';
import Radio from '../Radio';
import stateful from '../../../../.storybook/decorators/stateful';

storiesOf('Atoms / Forms / FieldGroup', module)
    .addDecorator(stateful())
    .add('radio buttons', () => (onChange, state) => (
        <FieldGroup name="test" value={ state.test } onChange={ onChange }>
            First
            {' '}
            <Radio value="first" />
            <br />
            Second
            {' '}
            <Radio value="second" />
            <br />
            Third
            {' '}
            <Radio value="third" />
        </FieldGroup>
    ))
    .add('checkboxes', () => (onChange, state) => (
        <FieldGroup name="test" value={ state.test } multiple onChange={ onChange }>
            First
            {' '}
            <Checkbox value="first" />
            <br />
            Second
            {' '}
            <Checkbox value="second" />
            <br />
            Third
            {' '}
            <Checkbox value="third" />
        </FieldGroup>
    ));
