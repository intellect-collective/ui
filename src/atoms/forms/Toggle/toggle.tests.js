import React from 'react';
import Toggle from '.';
import Form from '../Form';
import FieldGroup from '../FieldGroup';
import {
    basicChangeableTests,
    rerenderSuppressionTests
} from '../../../../test/utils';

const validProps = {
    name: 'test',
    value: 'fire',
    onChange: () => {}
};

const expectedProps = {
    name: 'test',
    value: 'fire',
    onChange: expect.any(Function)
};

describe('Toggle', () => {
    basicChangeableTests(Toggle, 'input', validProps, expectedProps);
    rerenderSuppressionTests(Toggle, validProps);

    describe('Form Context Usage', () => {
        it('Gets checked state from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Toggle name="test" value="sausages" />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(true);
        });

        it('Prefers local attributes over form attributes', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <Toggle name="test" value="sausages" checked={ false } />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Remains unchecked if form field has incorrect value', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'hotdogs' }}>
                    <Toggle name="test" value="sausages" />
                </Form>
            );
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });

        it('Can check multiple', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: ['hotdogs', 'sausages'] }}>
                    <Toggle name="test" value="hotdogs" />
                    <Toggle name="test" value="sausages" />
                    <Toggle name="test" value="burgers" />
                </Form>
            );
            expect(wrapper.find('input[value="hotdogs"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="sausages"]').prop('checked')).toEqual(true);
            expect(wrapper.find('input[value="burgers"]').prop('checked')).toEqual(false);
        });
    });

    describe('FieldGroup Context Usage', () => {
        it('Gets name from group', () => {
            const wrapper = mount(
                <FieldGroup name="test">
                    <Toggle value="sausages" />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
        });

        it('Prefers local name over group name', () => {
            const wrapper = mount(
                <FieldGroup name="test">
                    <Toggle name="test2" value="sausages" />
                </FieldGroup>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test2');
        });
    });

    describe('Form + FieldGroup Context Usage', () => {
        it('Gets state from form', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <FieldGroup name="test">
                        <Toggle value="sausages" />
                    </FieldGroup>
                </Form>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(true);
        });

        it('Prefers state from field', () => {
            const wrapper = mount(
                <Form action="/" data={{ test: 'sausages' }}>
                    <FieldGroup name="test">
                        <Toggle value="sausages" checked={ false } />
                    </FieldGroup>
                </Form>
            );
            expect(wrapper.find('input').prop('name')).toEqual('test');
            expect(wrapper.find('input').prop('checked')).toEqual(false);
        });
    });
});
