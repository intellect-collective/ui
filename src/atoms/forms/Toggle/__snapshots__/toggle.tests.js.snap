// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Should accept valid props 1`] = `
<GroupContext
  name="test"
  onChange={[Function]}
  value="fire"
>
  <FormContext
    name="test"
    onChange={[Function]}
    value="fire"
  >
    <Changeable
      name="test"
      onChange={[Function]}
      value="fire"
    >
      <Toggle
        name="test"
        onChange={[Function]}
        value="fire"
      >
        <div
          className="toggle"
        >
          <input
            id="test"
            name="test"
            onChange={[Function]}
            type="checkbox"
            value="fire"
          />
          <label
            htmlFor="test"
          >
            <span
              className="switch"
            >
              <span
                className="switch-inner"
                tabIndex={-1}
              />
            </span>
          </label>
        </div>
      </Toggle>
    </Changeable>
  </FormContext>
</GroupContext>
`;
